src/Compilers/Core/Portable/NativePdbWriter/ComMemoryStream.cs

```diff
@@ -52,6 +63,32 @@ public void CopyTo(Stream stream)
             }
         }
 
+        public IEnumerable<ArraySegment<byte>> GetChunks()
+        {
+            int chunkIndex = 0;
+            int remainingBytes = _length;
+            while (remainingBytes > 0)
+            {
+                int bytesToCopy;
+
+                byte[] chunk;
+                if (chunkIndex < _chunks.Count)
+                {
+                    bytesToCopy = Math.Min(_chunkSize, remainingBytes);
+                    chunk = _chunks[chunkIndex++];
+                }
+                else
+                {
+                    chunk = new byte[remainingBytes];
```

Comment:
>chunk = new byte[remainingBytes]; [](start = 20, length = 33)

It looks like this array could be arbitrarily big. Would it make sense to limit it size to ```_chunkSize``` and reuse the same array for all trailing empty chunks?

Reply:
I did that in previous version of the PR. Jared pointed out that since the method returns writable array, the underlying caching strategy would leak to the caller if we reused the chunk. 

The zeroed array is only created when the caller seeks behind the end of the stream and doesn't write there. That doesn't happen in practice (PDBs don't end with many 0 bytes).

---
In reply to: [137344715](https://github.com/dotnet/roslyn/pull/21712#discussion_r137344715) [](ancestors = 137344715)

Reply:
I'll add a comment to clarify.

---
In reply to: [137378003](https://github.com/dotnet/roslyn/pull/21712#discussion_r137378003) [](ancestors = 137378003,137344715)

Reply:
> the underlying caching strategy would leak to the caller if we reused the chunk.

It is still leaking. Changes to this array are not reflected in the stream. I think it would be better o not expose data in this form. We could simply add an API that calculates hash, if that is what we are using it for.

Reply:
That's a good point. However, I think we should instead change it to add the zeroed array into the stream. In future I'd like to move the PDB writer to Microsoft.DiaSymreader and that can't depend on crypto APIs. I'll file a bug to follow up.

Reply:
https://github.com/dotnet/roslyn/issues/21951